service/tbc/crawler.go

```diff
@@ -274,92 +361,142 @@ func processTxs(cp *chaincfg.Params, blockHash *chainhash.Hash, txs []*btcutil.T
+			log.Infof("Tx indexer: %v tx cache %v%%", hh, cp)
 		}
 		if cp > txsPercentage {
 			// Set txsMax to the largest tx capacity seen
 			s.cfg.MaxCachedTxs = max(len(txs), s.cfg.MaxCachedTxs)
+			last = hh
 			// Flush
 			break
 		}
 
-		// If set we may have to exit early
-		if circuitBreaker {
-			if height >= maxHeight-1 {
+		// Exit if we processed the provided end hash
+		if endHash.IsEqual(&hash) {
+			last = hh
+			break
+		}
+
+		// Move to next block
+		height := bh.Height + 1
+		bhs, err := s.db.BlockHeadersByHeight(ctx, height)
+		if err != nil {
+			if errors.Is(err, database.ErrNotFound) {
+				log.Infof("No more blocks at: %v", height)
 				break
 			}
+			return 0, last, fmt.Errorf("block headers by height %v: %w",
+				height, err)
+		}
+		if len(bhs) > 1 {
+			panic("FIXME handle multiple block headers") // XXX: Handle correctly
 		}
+		// Verify it connects to parent
+		if !hash.IsEqual(bhs[0].ParentHash()) {
+			return 0, last, fmt.Errorf("%v does not connect to: %v",
+				bhs[0], hash)
+		}
+		hh.Hash = *bhs[0].BlockHash()
+		hh.Height = bhs[0].Height
 	}
 
-	return blocksProcessed, nil
+	return blocksProcessed, last, nil
 }
 
-// TxIndexer starts indexing at start height for count blocks. If count is 0
-// the indexers will index to tip. It does NOT verify that the provided start
-// height is correct. This is the version of the function that has no training
-// wheels and is meant for internal use only.
-func (s *Server) TxIndexer(ctx context.Context, height, count uint64) error {
+func (s *Server) TxIndexer(ctx context.Context, endHash *chainhash.Hash) error {
 	log.Tracef("TxIndexer")
 	defer log.Tracef("TxIndexer exit")
 
-	var maxHeight uint64
-	circuitBreaker := false
-	if count != 0 {
-		circuitBreaker = true
-		maxHeight = height + count
+	// Verify exit condition hash
+	if endHash == nil {
+		return errors.New("must provide an end hash")
+	}
+	_, endHeight, err := s.BlockHeaderByHash(ctx, endHash)
+	if err != nil {
+		return fmt.Errorf("blockheader hash: %w", err)
+	}
+
+	// Verify start point is not after the end point
+	txHH, err := s.TxIndexHash(ctx)
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			return fmt.Errorf("tx indexer : %w", err)
+		}
+		txHH = &HashHeight{
+			Hash:   *s.chainParams.GenesisHash,
+			Height: 0,
+		}
 	}
 
+	// XXX we need training wheels here. We can't blind accept the end
+	// without asserting if it is either ihigher in the chain or is a
+	// forced for.
+	// XXX check cumulative? check fork?
+
 	// Allocate here so that we don't waste space when not indexing.
 	txs := make(map[tbcd.TxKey]*tbcd.TxValue, s.cfg.MaxCachedTxs)
-	// log.Infof("max %v %v", s.cfg.MaxCachedTxs, s.cfg.MaxCachedTxs*(105))
-	// return nil
 	defer clear(txs)
 
-	log.Infof("Start indexing transactions at height %v count %v", height, count)
+	log.Infof("Start indexing Txs at hash %v height %v", txHH.Hash, txHH.Height)
+	log.Infof("End indexing Txs at hash %v height %v", endHash, endHeight)
```

Comment:
Should these logs be at the `info` level? If they are primarily for debugging, I think the `debug` level would be better.

Reply:
I just typed a few sentence here explaining why but github ate them.

Basically, we need to know how long we will be running.

Reply:
See response: https://github.com/hemilabs/heminetwork/pull/152#discussion_r1650879693